Give an option to remove 'using namespace'

deepikabhavnani opened the pull request:
### Description
~~DO NOT MERGE~~

I have created this PR to get input from all teams for solution to namespace issue. 
Purpose:
1. To give an option to remove "using namespace" with `MBED_NO_NAMESPACE` macro
2. Move all classes in Mbed OS inside `mbed` namespace. Addition of existing classes inside mbed namespace breaks compatibility, hence below code should be added to maintain compatibility.

```
// Added "using" for backwards compatibility
#ifndef MBED_NO_NAMESPACE
using mbed::<Classname>;
#endif
```

List of all classes to be moved to `mbed` namespace. Feel free to add to the list or comment below if you know of any class additions.

[ ] All Block Devices
[ ] Filesystem's
 
Related: https://github.com/ARMmbed/mbed-os/issues/6684

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Feature
    [X] Breaking change

CC @geky @bridadan @dlfryar @0xc0170 @pan- @kjbracey-arm

kjbracey commented:
Looks conceptually good to me. Is the intent then that application code would either define this `MBED_NO_NAMESPACE` globally or on a per-file basis before including mbed header files?

Little bit doubtful about the name - it's more `MBED_NO_USING`, but then that sounds weird. `MBED_NO_GLOBAL_NAMESPACE`?

There is another possible alternative approach. You could do what the C++ library does and make it so that `#include <mbed>` doesn't do the `using`, but `#include <mbed.h>` does.

That is a pattern I've not seen outside the standard C++ library, but given that it's there it shouldn't be totally unfamiliar. But it doesn't match our use of `.h` for all other header files, and not sure what other problems lack of `.h` would cause.

Maybe `#include` of some other name that omits the `using` - `"mbed_os.h"`?  That approach doesn't cover backwards compatibility with people directly including `ObservingBlockDevice.h` though.

pan- commented:
> There is another possible alternative approach. You could do what the C++ library does and make it so that #include <mbed> doesn't do the using, but #include <mbed.h> does.

I like the idea but I fear it confuses some of our developers. Our codebase itself is not consistent with the practice.

For the name, like @kjbracey-arm I think it should be changed. `MBED_NO_NAMESPACE ` feels like it removes namespace. I can think of few names: `MBED_NO_GLOBAL_USING_DIRECTIVE` or `MBED_NO_GLOBAL_USING_NAMESPACE`. 

What's the path going forward ? Is there a plan to set this macro by default ? It may be worthwhile to think of how components are mapped into namespaces as we may require more than just the mbed namespace.

deepikabhavnani commented:
> Is the intent then that application code would either define this MBED_NO_NAMESPACE globally or on a per-file basis before including mbed header files?

Can leave that to application. It should work both ways, if we make sure mbed-os compiles with flag enabled globally. Default we cannot enable it, but adding it to CI or not will be question. 

> For the name, like @kjbracey-arm I think it should be changed. `MBED_NO_NAMESPACE` feels like it removes namespace. I can think of few names:`MBED_NO_GLOBAL_USING_DIRECTIVE` or `MBED_NO_GLOBAL_USING_NAMESPACE`.

Yes , good names are appreciated.  `MBED_NO_GLOBAL_USING_NAMESPACE` / `MBED_NO_GLOBAL_USING_DIRECTIVE` 👍 

> #include <mbed> doesn't do the using, but #include <mbed.h> does.

Agree with @pan- it will be confusing, Also as you mentioned it will not be backward compatible we should avoid that solution.

bmcdonnell-ionx commented:
As a user, I agree with not dropping the `.h` extension from your header file names. AFAIK it's not a convention outside the C++ library.

> Default we cannot enable it

How about for Mbed OS 6.0?

Or, zooming out, why is `mbed.h` necessary or desirable at all? For instance, you don't `#include <TheWholeCppLibrary.h>`; you `#include` just the headers you need. Why is it a good thing to have a single header that makes the entire OS/lib interface visible?

@pan- said previously (https://github.com/ARMmbed/mbed-os/issues/5679#issuecomment-355995183):

> From my perspective, the issue in that case is the use of `mbed.h` directly in os/production code. It is not designed for that usage and unsafe regards to name collision by _design_. The sole and only viable solution is to not use it in code that needs to be name collision free.
>
> Unfortunately uses of `mbed.h` are not properly documented which lead developers to use it in the wrong context. That would be nice to have a clear documentation bloc explaining what `mbed.h` is as well as how and where it should be used.
>
> About the changes proposed, I personally believe that even if `mbed-os.h` solves the name collision issue it is not a facility that should be used in production code except if we want to encourage build time increases.

kjbracey commented:
> Or, zooming out, why is mbed.h necessary or desirable at all? 

My understanding is that it's desirable for "simplicity" - simple apps just include "mbed.h". Same basic argument as putting every single folder on the include path, and putting everything in the global namespace.

Also, possibly some of the core built-in stuff doesn't have as obviously clear "what to include" rules as "Thread.h" for `Thread`, so I guess it mops up a bunch of miscellaneous stuff - types, error codes etc, giving us leeway to restructure a bit. 

I would say it should not be used in any OS or library header files, so that the name pollution isn't forced on application code.

0xc0170 commented on events/mbed_shared_queues.cpp:
```diff
@@ -15,7 +15,10 @@
 #include "events/mbed_shared_queues.h"
-#include "mbed.h"
+
+#ifdef MBED_CONF_RTOS_PRESENT
```
I can see similar changes in this PR that could be separated from adding new config (changes in events files at least)? As I expect with this config change to touch only mbed or std namespace (add it ,remove it).

deepikabhavnani replied:
Removing namespace i.e building mbed-os with "-DMBED_NO_GLOBAL_USING_DIRECTIVE" results in build failures as 'mbed.h' added here will not add RTOS / Mbed namespace in cpp files as well. 

This cleanup "i.e. add only required namespace/header file" is needed for removal of namespace to work. I can add it as separate PR, and make this PR to be dependent on that.

bmcdonnell-ionx commented:
@kjbracey-arm,

> I would say it should not be used in any OS or library header files, so that the name pollution isn't forced on application code.

I absolutely agree.

Assuming you do keep it, please also document in appropriate places (when it becomes true) that application code doesn't *need* to `#include "mbed.h"`.

deepikabhavnani commented:
> I would say it should not be used in any OS or library header files, so that the name pollution isn't forced on application code.

👍 Removing mbed.h here - https://github.com/ARMmbed/mbed-os/pull/7864

deepikabhavnani commented:
Dependent on https://github.com/ARMmbed/mbed-os/pull/7864 https://github.com/ARMmbed/mbed-os/pull/8002 https://github.com/ARMmbed/mbed-os/pull/8008

cmonr commented:
https://github.com/ARMmbed/mbed-os/pull/7760#issuecomment-429056526

@deepikabhavnani The PRs that were referenced in this comment,
1) They're all needed before this PR can progress, correct?
2) Should all of those PRs be released at the same time/have the same "release version" label?

deepikabhavnani commented:
> They're all needed before this PR can progress, correct?

Yes

> Should all of those PRs be released at the same time/have the same "release version" label?

Not really needed to have same release version, as previous PR's are just cleanup. But all namespace related changes should be in same version. For now we have this PR and https://github.com/ARMmbed/mbed-os/pull/8329

cmonr commented:
@deepikabhavnani Might want to think about rebasing this now before #7864 is in.

deepikabhavnani commented:
This is dependent on 7864, will rebase after that is merged

cmonr commented:
@deepikabhavnani Rebase at will!

Commit:
Add BlockDevice and Filesystem classes inside mbed namespace.

Adding new modules inside the namespace could be breaking change for existing code base
hence add `using namespace::class` for classes newly added to mbed namespace to maintian
backwards compatibility.

MBED_NO_GLOBAL_USING_DIRECTIVE is added to remove auto-addition of namespace
Macro guard `MBED_NO_GLOBAL_USING_DIRECTIVE` is added around namespace, to avoid
polluting users namespace.

Commit:
MBED_NO_GLOBAL_USING_DIRECTIVE is added to remove auto-addition of namespace

Macro guard `MBED_NO_GLOBAL_USING_DIRECTIVE` is added around namespace, to avoid
polluting users namespace.

deepikabhavnani commented:
@ARMmbed/mbed-os-storage  - Filesystem + blockdevice is moved to mbed namespace. Before adding `nvstore` and  `system_storage` to mbed namespace, wanted to know if we have any concerns or restrictions? Also for now I am not sure of `FEATURE_STORAGE`

deepikabhavnani commented:
Rebased to resolve conflict

@ARMmbed/mbed-os-storage @pan- @geky - Please review

dannybenor commented:
need a few days before review

deepikabhavnani commented:
> need a few days before review

@dannybenor - Apart from review please let us know if any other module must be added to `mbed` namespace

adbridge commented:
@deepikabhavnani Could you please rebase this?

deepikabhavnani commented:
@adbridge - Will wait for storage changes for 5.11 to be merged, also seems this will not be reviewed till than.

bulislaw reviewed (approved):
+1

0xc0170 commented:
@deepikabhavnani All storage PR landed, or still waiting for some here?

deepikabhavnani commented:
> @deepikabhavnani All storage PR landed, or still waiting for some here? 

Review from storage team is pending as well

dannybenor commented:
@offirko can you please review?

offirko reviewed (approved):
no remarks, LGTM.

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

deepikabhavnani commented:
Rebased

cmonr commented:
CI started.

deepikabhavnani commented:
@cmonr  - Fixed for failing device. Needs CI restart

cmonr commented:
https://github.com/ARMmbed/mbed-os/pull/7760#issuecomment-441769253

I like it. So fast, didn't even realize this had a problem.

CI restarted.
